internal/training/grpo_engine/kernels.c

```
@@ -0,0 +1,764 @@
+    /* qs starts at byte 16 (after d=2 + dmin=2 + scales=12) */
+    const uint8_t *qs = block + 16;
+    uint8_t q;
+    if (idx % 2 == 0)
+        q = qs[idx / 2] & 0x0F;
+    else
+        q = qs[idx / 2] >> 4;
+
+    return d * (float)q - dmin;
+}
+
+void grpo_matmul_q4(float *out, const float *x, const void *W_packed,
```

Cppcheck (2.21.0)

[style] The function 'grpo_matmul_q4' should have static linkage since it is not used outside of its translation unit.

(staticFunction)